New SetRenderBufferMask StateChange

vampcat opened the pull request:
This PR
- Adds a new SetRenderBufferMask StateChange
- Moves MarkerTask to end of file, so it's no longer the first class in file.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

emanuele3d commented:
@vampcat: I'm happy with the changes. Can you please write a javadoc for the SetRenderBufferMask class? Also please check its public methods to verify if they need javadocs too.

vampcat commented:
Aye aye cap'n!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetRenderBufferMask.java:
```diff
@@ -24,7 +24,10 @@
+ * This StateChange generates the tasks that set and reset render masks.
+ *
+ * The rendering pipeline can either write to the Color buffer, the DepthStencil buffer
+ * or the Light buffer, all of which can be toggled independently of each other.
```
Ok, two concepts have to emerge here:
1. the render mask is part of the state of an FBO - not the opengl state.
2. this particular StateChange really only works on one of the read/write only FBOs - it doesn't work with any other FBO.

vampcat replied:
Why does it only work on read/write FBOs ?

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetRenderBufferMask.java:
```diff
@@ -37,6 +40,15 @@
+     * @param renderToColorBuffer A boolean indicating whether the Color buffer of the given FBO should be written to.
+     * @param renderToDepthBuffer A boolean indicating whether the DepthStencil buffer of the given FBO should be written to.
+     * @param renderToLightBuffer A boolean indicating whether the Light buffer of the given FBO should be written to.
+     */
```
You might want to revisit this javadoc too in light of the previous comment: this class is actually very specialized.

vampcat commented:
I believe that should take care of it all, feel free to correct me.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/BackdropNode.java:
```diff
@@ -90,6 +91,7 @@ public void initialise() {
         new WireframeTrigger(renderingDebugConfig, this);
 
         addDesiredStateChange(new BindFBO(READONLY_GBUFFER, displayResolutionDependentFBOs));
+        addDesiredStateChange(new SetRenderBufferMask(READONLY_GBUFFER, displayResolutionDependentFBOs, true, false, false));
```
Can we change the signature here so that the mask (true, false false) is in the first three arguments of the constructor?

I'm also wondering if we should rename SetRenderBufferMask to SetFboWriteMask instead.

vampcat replied:
The reason I chose this name was, the function called internally is fbo.setRenderBufferMask. I can change the name regardless, if you want.

Also, I kept it this way because it's data (fbo), and the settings for the data (true, false, false). But I'll change it :D

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetRenderBufferMask.java:
```diff
@@ -0,0 +1,152 @@
+ * This StateChange modifies the FBO to mask or unmask these three buffers, independent of each other.
+ * Note that this StateChange assumes the given FBO has the required attachments, and trying to unmask the
+ * depth buffer or light buffer for a normal FBO with default configuration -might- lead to undefined behaviour.
+ */
```
I'd write:

```java
/**
* Instances of this class generate task setting and resetting an FBO's write mask. 
*
* A write mask is useful to render to an FBO leaving some of its attachments untouched. 
* 
* This particular state change independently enables/disables writing to the color, depth and light accumulation 
* attachments of an FBO. At this stage this functionality makes sense only in the context of the readOnly/writeOnly 
* gBuffers as only those buffers have all the attachments mentioned.
*
* The behaviour of this state change in relation to FBOs that do not have all the relevant attachments has not been 
* investigated.
*/
```

vampcat replied:
I assure you, I'll learn the art of writing javadocs soon enough xD

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetRenderBufferMask.java:
```diff
@@ -0,0 +1,152 @@
+     * @param fboManager The FBOManager responsible for managing the given FBO.
+     * @param renderToColorBuffer A boolean indicating whether the Color buffer of the given FBO should be written to.
+     * @param renderToDepthBuffer A boolean indicating whether the DepthStencil buffer of the given FBO should be written to.
+     * @param renderToLightBuffer A boolean indicating whether the Light buffer of the given FBO should be written to.
```
I'd write "the Light _accumulation_ buffer".

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetRenderBufferMask.java:
```diff
@@ -0,0 +1,152 @@
+    /**
+     * Creates an instance of this class with the given parameters.
+     *
+     * @param fboName A ResourceUrn identifying the FBO whose render masks have to be modified.
```
I'd write:

`@param fboName A ResourceUrn identifying the FBO whose render masks have to be modified - usually only the writeOnlyGBuffer FBO. `

In practice right now this state change is used also for the read-only gBuffer, but that's an aberration that will be corrected eventually.

vampcat commented:
I think I have addressed all the issues in the last commit, while also removing a few FBOs that are now redundant (as I mentioned, we won't need any Node to subscribe to FBOManagers very soon. I'm trying my best to make it happen!).

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

emanuele3d commented:
@Cervator: another good to go PR. Thank you @vampcat, especially for the latest changes.

emanuele3d commented:
@Cervator: shall we have a "Ready to Merge" step in the github pipeline?

Cervator commented:
IMHO ready to merge is more or less "Changes Approved" + all status checks have passed :-)

As part of merging naturally we'd want to do some added testing which may find more stuff.

Actually we should probably try to be more exact with the review status - often commits are added that address the review, but we don't bother clicking the approve button instead just commenting "Looks good!", so it doesn't show "Changes Approved" on the main PR page.

In any case I checked this out and am merging! :-)
